addon/helpers/join.js

```diff
@@ -1,20 +1,13 @@
+import { helper } from '@ember/component/helper';
 import { isArray as isEmberArray } from '@ember/array';
-import Helper from '@ember/component/helper';
-import { observer } from '@ember/object';
-import { set } from '@ember/object';
 
-export default Helper.extend({
-  compute([separator, array]) {
-    if (isEmberArray(separator)) {
-      array = separator;
-      separator = ',';
-    }
+function join([separator, array]) {
```

Comment:
Perhaps we should also switch the order of these so that the optional argument goes at the end? 

```hbs
{{join this.items}}
{{join this.items "-"}}
```

If we like, we can emit a deprecation warning if a string is passed as the first argument

Reply:
It might be worth considering the API for all these helpers, eg:

perhaps:

`(map-by "fullName" users)`

should be?

`(map-by users "fullName)`

Reply:
I now see that the pattern of providing the array as the last argument is followed throughout the addon. I think we should leave the APIs as is in this PR and have a discussion on some future API in an issue. I'll open one soon

Reply:
> For all non-unary helpers, the subject of the helper function will always be the last argument. This way the arguments are better readable if you compose together multiple helpers

ah

Reply:
@GavinJoyce FWIW that pattern (stolen from FP languages) is *sooooo* 😍 once you get used to it. It's precisely what makes these *composable*: you can partially apply them (historically with `action`, now with `fn`) and have the next one *also* partially applied, and so on… and only supply the relevant data at the very end.

Reply:
Thanks. Now that I see how it composes, I'm a fan.